components/debounce/browser/debounce_download_service.cc

```diff
@@ -0,0 +1,216 @@
+  converter->RegisterCustomValueField<extensions::URLPatternSet>(
+      kExclude, &DebounceRule::exclude_pattern_set_, GetURLPatternSetFromValue);
+  converter->RegisterCustomField<DebounceAction>(
+      kAction, &DebounceRule::action_, &ParseDebounceAction);
+  converter->RegisterStringField(kParam, &DebounceRule::param_);
+}
+
+bool DebounceRule::Apply(const GURL& original_url, GURL* final_url) {
+  // If URL matches an explicitly excluded pattern, this rule does not apply.
+  if (exclude_pattern_set_.MatchesURL(original_url))
+    return false;
+  // If URL does not match an explicitly included pattern, this rule does not
+  // apply.
+  if (!include_pattern_set_.MatchesURL(original_url))
+    return false;
+
+  if (action_ == kDebounceRedirectToParam ||
+      action_ == kDebounceBase64DecodeAndRedirectToParam) {
+    std::string unescaped_value;
+    if (!net::GetValueForKeyInQuery(original_url, param_, &unescaped_value))
+      return false;
+    GURL new_url;
+    if (action_ == kDebounceBase64DecodeAndRedirectToParam) {
+      std::string base64_decoded_value;
+      if (!base::Base64UrlDecode(unescaped_value,
+                                 base::Base64UrlDecodePolicy::IGNORE_PADDING,
+                                 &base64_decoded_value))
+        return false;
+      new_url = GURL(base64_decoded_value);
+    } else {
+      new_url = GURL(unescaped_value);
+    }
+
+    // Failsafe: ensure we got a valid URL out of the param.
+    if (!new_url.is_valid() || !new_url.SchemeIsHTTPOrHTTPS())
+      return false;
+
+    // Failsafe: never redirect to the same site.
+    url::Origin original_origin = url::Origin::Create(original_url);
+    url::Origin debounced_origin = url::Origin::Create(new_url);
+    if (original_origin.IsSameOriginWith(debounced_origin))
+      return false;
+
+    *final_url = new_url;
+    return true;
+  }
+
+  // Unknown actions always return false, to allow for future updates to the
+  // rules file which may be pushed to users before a new version of the code
+  // that parses it.
+  return false;
+}
+
+DebounceDownloadService::DebounceDownloadService(
+    LocalDataFilesService* local_data_files_service)
+    : LocalDataFilesObserver(local_data_files_service), weak_factory_(this) {}
+
+DebounceDownloadService::~DebounceDownloadService() {}
+
+void DebounceDownloadService::LoadDirectlyFromResourcePath() {
+  base::FilePath dat_file_path = resource_dir_.AppendASCII(kDebounceConfigFile);
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
+      base::BindOnce(&brave_component_updater::GetDATFileAsString,
+                     dat_file_path),
+      base::BindOnce(&DebounceDownloadService::OnDATFileDataReady,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void DebounceDownloadService::OnDATFileDataReady(std::string contents) {
+  rules_.clear();
+  host_cache_.clear();
+  if (contents.empty()) {
+    LOG(ERROR) << "Could not obtain debounce configuration";
```

Comment:
should we be clearing the old data if the download fails for some reason? Seems like we'd rather have something than nothing?

Reply:
also this should probably be a VLOG